airbyte-integrations/connectors/source-mixpanel/source_mixpanel/source.py

```diff
@@ -0,0 +1,839 @@
+        funnel_id = int(params["funnel_id"][0])
+
+        # read and transform records
+        records = response.json().get(self.data_field, {})
+        for date_entry in records:
+            # for each record add funnel_id, name
+            yield {
+                "funnel_id": funnel_id,
+                "name": self.funnels[funnel_id],
+                "date": date_entry,
+                **records[date_entry],
+            }
+
+    def get_updated_state(
+        self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]
+    ) -> Mapping[str, Mapping[str, str]]:
+        """Update existing stream state for particular funnel_id
+        stream_state = {
+            'funnel_id1_str' = {'date': 'datetime_string1'},
+            'funnel_id2_str' = {'date': 'datetime_string2'},
+             ...
+            'funnel_idX_str' = {'date': 'datetime_stringX'},
+        }
+        NOTE: funnel_id1 type:
+            - int in latest_record
+            - str in current_stream_state
+        """
+        funnel_id: str = str(latest_record["funnel_id"])
+
+        latest_record_date: str = latest_record.get(self.cursor_field, str(self.start_date))
+        stream_state_date: str = str(self.start_date)
+        if current_stream_state and funnel_id in current_stream_state:
+            stream_state_date = current_stream_state[funnel_id]["date"]
+
+        # update existing stream state
+        current_stream_state[funnel_id] = {"date": max(latest_record_date, stream_state_date)}
+
+        return current_stream_state
+
+
+class EngageSchema(MixpanelStream):
+    """Engage helper stream for dynamic schema extraction"""
+
+    primary_key = None
+    data_field = "results"
+
+    def path(self, **kwargs) -> str:
+        return "engage/properties"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        response.json() example:
+        {
+            "results": {
+                "$browser": {
+                    "count": 124,
+                    "type": "string"
+                },
+                "$browser_version": {
+                    "count": 124,
+                    "type": "string"
+                },
+                ...
+                "_some_custom_property": {
+                    "count": 124,
+                    "type": "string"
+                }
+            }
+        }
+        """
+        records = response.json().get(self.data_field, {})
+        for property_name in records:
+            yield {
+                "name": property_name,
+                "type": records[property_name]["type"],
+            }
+
+
+class Engage(MixpanelStream):
+    """Return list of all users
+    API Docs: https://developer.mixpanel.com/reference/engage
+    Endpoint: https://mixpanel.com/api/2.0/engage
+    """
+
+    http_method = "POST"
+    data_field = "results"
+    primary_key = "distinct_id"
+    page_size = 1000  # min 100
+    _total = None
```

Comment:
what does `_total` represent?

Reply:
it is used for paging. We need to store a total number of results that exist only on first page.